docs/design-proposals/2022-04 Parquet.md

```diff
@@ -0,0 +1,256 @@
+
+```
+repeated group Attrs {
+    required binary  Key (STRING);
+
+    # Only one of these will be set
+    optional binary  Value (STRING);
+    optional boolean ValueBool;
+    optional double  ValueDouble;
+    optional int64   ValueInt (INT(64,true));
+    optional binary  ValueArray (STRING);
+    optional binary  ValueKVList (STRING);
+}
+```
+
+Pros:
+
+* Preserves the functionality of parquet column min/max statistics which can be leveraged during search
+* Native data types should be more efficient to store and process
+
+Cons:
+
+* Many nulls - as each attribute only populates 1 column, the other 5 are guaranteed to be null. This has a non-trivial increase in storage size.
+
+### Event Attributes
+Span event attributes are stored as JSON-encoded strings in a generic key/value map. This is by far the most space-efficient encoding and the trade-off of decreased searchability seems worthwhile.  Storing event attributes this way reduces the block size ~16% for our dataset, which is huge.  There are currently no use cases to search event attributes, but we can revisit this in the future if needed.
+
+```
+repeated group Attrs {
+    required binary Key (STRING);
+    required binary Value (STRING);
+}
+```
+
+### Compression and Encoding
+Parquet has robust support for many compression algorithms and data encodings. We have found excellent combinations of storage size and performance with the following:
+
+1. LZ4 Compression - Enable on all columns
+2. Dictionary encoding - Enable on all string columns (including byte array ParentSpanID). Most strings are very repetitive so this works well to optimize storage size.  However we can greatly speed up search by inspecting the dictionary first and eliminating pages with no matches.
+3. Time and duration unix nanos - Delta encoding
+4. Rarely used columns such as DroppedAttributesCount - These columns are usually all zeroes, RLE works well.
+
+### Bloom Filters
+Parquet has native support for bloom filters however we are not using them at this time.  Tempo already has sophisticated support for sharding and caching bloom filters, and we will continue to leverage that for now.
+
+## Results from Local Testing
+Here are some interesting column sizes from a block of size ~600MB containing ~150K traces:
+
+```
+column.Trace.DurationNanos                                                    154414 values  0.64 MB
+column.Trace.ResourceSpans.Resource.ServiceName                              2672245 values  2.01 MB
+column.Trace.ResourceSpans.InstrumentationLibrarySpans.Spans.HttpStatusCode 10428415 values  6.35 MB 
+column.Trace.ResourceSpans.InstrumentationLibrarySpans.Spans.ID             10428415 values 82.37 MB  # this is never used!
+```
+
+Some super early benchmarks on local SSDs:
+
+* Searching protobuf data for a simple query `cluster=ops and minDuration=1s`
+
+```
+=== RUN   TestSearchProto
+Traces : 55
```

Comment:
I think "Traces" means "Traces Matched"?

Reply:
Yes. Correct. We added that metric in our local tests to confirm both proto/parquet are returning the same number of hits. We can leave this line out in this doc though.